Ingress Controller 0.7.1 Crashes when Kubernetes Secret is not formatted as expected

### Summary
Kong Ingress Controller Crashes when Kong Credential secret does not contain key `key`

**Kong Ingress controller version**
0.7.1

**Kong or Kong Enterprise version**
1.4.3

### Kubernetes version
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-13T11:51:44Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.9-eks-c0eccc", GitCommit:"c0eccca51d7500bb03b2f163dd8d534ffeb2f7a2", GitTreeState:"clean", BuildDate:"2019-12-22T23:14:11Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}

### Environment
AWS EKS

### What happened
Kong Ingress Controller Crashes trying to read in a secret that does not have a key in the secret named "key"

### Expected behavior
Kong Ingress Controller will gracefully handle this without null pointer exception resulting in crash

### Steps To Reproduce

1. Create a secret with key `kongCredType` and do not create a subsequent key named `key`

### Output
```
E0207 19:49:04.782376       1 runtime.go:73] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 277 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x142a600, 0x2229630)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/runtime/runtime.go:69 +0x7b
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/runtime/runtime.go:51 +0x82
panic(0x142a600, 0x2229630)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/hbagdi/deck/file.(*stateBuilder).ingestKeyAuths(0xc000ec99f0, 0xc00032b0a0, 0x2, 0x2, 0x2, 0xc00032b0a0)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/builder.go:222 +0x408
github.com/hbagdi/deck/file.(*stateBuilder).consumers(0xc000ec99f0)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/builder.go:161 +0x57b
github.com/hbagdi/deck/file.(*stateBuilder).build(0xc000ec99f0, 0x0, 0x0, 0x4)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/builder.go:56 +0x124
github.com/hbagdi/deck/file.Get(0xc0009e00e0, 0xc000049200, 0x1, 0x4, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/reader.go:51 +0x130
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).onUpdateDBMode(0xc0009a6000, 0xc0009e00e0, 0x13c6960, 0xc000c7a8e0)
	/kong-ingress-controller/internal/ingress/controller/kong.go:161 +0x1c4
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).OnUpdate(0xc0009a6000, 0xc000650a80, 0x0, 0x0)
	/kong-ingress-controller/internal/ingress/controller/kong.go:63 +0x26c
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).syncIngress(0xc0009a6000, 0x149a360, 0xc000156820, 0xc017934d78, 0x2adca41b2681c)
	/kong-ingress-controller/internal/ingress/controller/controller.go:117 +0x1fa
github.com/kong/kubernetes-ingress-controller/internal/ingress/task.(*Queue).worker(0xc00046fce0)
	/kong-ingress-controller/internal/ingress/task/queue.go:111 +0x2df
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0005407a8)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000ec9fa8, 0x3b9aca00, 0x0, 0x1, 0xc00051e2a0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:88
github.com/kong/kubernetes-ingress-controller/internal/ingress/task.(*Queue).Run(0xc00046fce0, 0x3b9aca00, 0xc00051e2a0)
	/kong-ingress-controller/internal/ingress/task/queue.go:59 +0x6b
created by github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).Start
	/kong-ingress-controller/internal/ingress/controller/controller.go:232 +0xe8
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x11ab8c8]
goroutine 277 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/runtime/runtime.go:58 +0x105
panic(0x142a600, 0x2229630)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/hbagdi/deck/file.(*stateBuilder).ingestKeyAuths(0xc000ec99f0, 0xc00032b0a0, 0x2, 0x2, 0x2, 0xc00032b0a0)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/builder.go:222 +0x408
github.com/hbagdi/deck/file.(*stateBuilder).consumers(0xc000ec99f0)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/builder.go:161 +0x57b
github.com/hbagdi/deck/file.(*stateBuilder).build(0xc000ec99f0, 0x0, 0x0, 0x4)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/builder.go:56 +0x124
github.com/hbagdi/deck/file.Get(0xc0009e00e0, 0xc000049200, 0x1, 0x4, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/pkg/mod/github.com/hbagdi/deck@v0.7.1-0.20191223190449-3d9f90945d9d/file/reader.go:51 +0x130
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).onUpdateDBMode(0xc0009a6000, 0xc0009e00e0, 0x13c6960, 0xc000c7a8e0)
	/kong-ingress-controller/internal/ingress/controller/kong.go:161 +0x1c4
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).OnUpdate(0xc0009a6000, 0xc000650a80, 0x0, 0x0)
	/kong-ingress-controller/internal/ingress/controller/kong.go:63 +0x26c
github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).syncIngress(0xc0009a6000, 0x149a360, 0xc000156820, 0xc017934d78, 0x2adca41b2681c)
	/kong-ingress-controller/internal/ingress/controller/controller.go:117 +0x1fa
github.com/kong/kubernetes-ingress-controller/internal/ingress/task.(*Queue).worker(0xc00046fce0)
	/kong-ingress-controller/internal/ingress/task/queue.go:111 +0x2df
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0005407a8)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000ec9fa8, 0x3b9aca00, 0x0, 0x1, 0xc00051e2a0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/wait/wait.go:88
github.com/kong/kubernetes-ingress-controller/internal/ingress/task.(*Queue).Run(0xc00046fce0, 0x3b9aca00, 0xc00051e2a0)
	/kong-ingress-controller/internal/ingress/task/queue.go:59 +0x6b
created by github.com/kong/kubernetes-ingress-controller/internal/ingress/controller.(*KongController).Start
	/kong-ingress-controller/internal/ingress/controller/controller.go:232 +0xe8
```

--- a/internal/ingress/controller/parser/kongstate/consumer.go
+++ b/internal/ingress/controller/parser/kongstate/consumer.go
@@ -33,12 +33,24 @@ func (c *Consumer) SetCredential(log logrus.FieldLogger, credType string, credCo
 			return fmt.Errorf("failed to decode key-auth credential: %w", err)
 
 		}
+		// TODO we perform these validity checks here because passing credentials without these fields will panic deck
+		// later on. Ideally this should not be handled in the controller, but we cannot currently handle it elsewhere
+		// (i.e. in deck or go-kong) without entering a sync failure loop that cannot actually report the problem
+		// piece of configuration. if we can address those limitations, we should remove these checks.
+		// See https://github.com/Kong/deck/pull/223 and https://github.com/Kong/kubernetes-ingress-controller/issues/532
+		// for more discussion.
+		if cred.Key == nil {
+			return fmt.Errorf("key-auth for consumer %s is invalid: no key", *c.Username)
+		}
 		c.KeyAuths = append(c.KeyAuths, &cred)
 	case "basic-auth", "basicauth_credential":
 		var cred kong.BasicAuth
 		err := decodeCredential(credConfig, &cred)
 		if err != nil {
 			return fmt.Errorf("failed to decode basic-auth credential: %w", err)
+		}
+		if cred.Username == nil {
+			return fmt.Errorf("basic-auth for consumer %s is invalid: no username", *c.Username)
 		}
 		c.BasicAuths = append(c.BasicAuths, &cred)
 	case "hmac-auth", "hmacauth_credential":
@@ -47,12 +59,18 @@ func (c *Consumer) SetCredential(log logrus.FieldLogger, credType string, credCo
 		if err != nil {
 			return fmt.Errorf("failed to decode hmac-auth credential: %w", err)
 		}
+		if cred.Username == nil {
+			return fmt.Errorf("hmac-auth for consumer %s is invalid: no username", *c.Username)
+		}
 		c.HMACAuths = append(c.HMACAuths, &cred)
 	case "oauth2":
 		var cred kong.Oauth2Credential
 		err := decodeCredential(credConfig, &cred)
 		if err != nil {
 			return fmt.Errorf("failed to decode oauth2 credential: %w", err)
+		}
+		if cred.ClientID == nil {
+			return fmt.Errorf("oauth2 for consumer %s is invalid: no client_id", *c.Username)
 		}
 		c.Oauth2Creds = append(c.Oauth2Creds, &cred)
 	case "jwt", "jwt_secret":
@@ -70,12 +88,18 @@ func (c *Consumer) SetCredential(log logrus.FieldLogger, credType string, credCo
 		if cred.Algorithm == nil || *cred.Algorithm == "" {
 			cred.Algorithm = kong.String("HS256")
 		}
+		if cred.Key == nil {
+			return fmt.Errorf("jwt-auth for consumer %s is invalid: no key", *c.Username)
+		}
 		c.JWTAuths = append(c.JWTAuths, &cred)
 	case "acl":
 		var cred kong.ACLGroup
 		err := decodeCredential(credConfig, &cred)
 		if err != nil {
 			log.Errorf("failed to process ACL group: %v", err)
+		}
+		if cred.Group == nil {
+			return fmt.Errorf("acl for consumer %s is invalid: no group", *c.Username)
 		}
 		c.ACLGroups = append(c.ACLGroups, &cred)
 	default:

--- a/internal/ingress/controller/parser/kongstate/consumer_test.go
+++ b/internal/ingress/controller/parser/kongstate/consumer_test.go
@@ -50,19 +50,11 @@ func TestConsumer_SetCredential(t *testing.T) {
 		{
 			name: "key-auth without key",
 			args: args{
-				credType: "key-auth",
-				consumer: &Consumer{
-					Consumer: kong.Consumer{
-						Username: &username,
-					},
-				},
+				credType:   "key-auth",
+				consumer:   &Consumer{Consumer: kong.Consumer{Username: &username}},
 				credConfig: map[string]string{},
 			},
-			result: &Consumer{
-				Consumer: kong.Consumer{
-					Username: &username,
-				},
-			},
+			result:  &Consumer{Consumer: kong.Consumer{Username: &username}},
 			wantErr: true,
 		},
 		{
@@ -104,19 +96,11 @@ func TestConsumer_SetCredential(t *testing.T) {
 		{
 			name: "basic-auth without username",
 			args: args{
-				credType: "basic-auth",
-				consumer: &Consumer{
-					Consumer: kong.Consumer{
-						Username: &username,
-					},
-				},
+				credType:   "basic-auth",
+				consumer:   &Consumer{Consumer: kong.Consumer{Username: &username}},
 				credConfig: map[string]string{},
 			},
-			result: &Consumer{
-				Consumer: kong.Consumer{
-					Username: &username,
-				},
-			},
+			result:  &Consumer{Consumer: kong.Consumer{Username: &username}},
 			wantErr: true,
 		},
 		{
@@ -162,19 +146,11 @@ func TestConsumer_SetCredential(t *testing.T) {
 		{
 			name: "hmac-auth without username",
 			args: args{
-				credType: "hmac-auth",
-				consumer: &Consumer{
-					Consumer: kong.Consumer{
-						Username: &username,
-					},
-				},
+				credType:   "hmac-auth",
+				consumer:   &Consumer{Consumer: kong.Consumer{Username: &username}},
 				credConfig: map[string]string{},
 			},
-			result: &Consumer{
-				Consumer: kong.Consumer{
-					Username: &username,
-				},
-			},
+			result:  &Consumer{Consumer: kong.Consumer{Username: &username}},
 			wantErr: true,
 		},
 		{
@@ -224,19 +200,11 @@ func TestConsumer_SetCredential(t *testing.T) {
 		{
 			name: "oauth2 without client_id",
 			args: args{
-				credType: "oauth2",
-				consumer: &Consumer{
-					Consumer: kong.Consumer{
-						Username: &username,
-					},
-				},
+				credType:   "oauth2",
+				consumer:   &Consumer{Consumer: kong.Consumer{Username: &username}},
 				credConfig: map[string]string{},
 			},
-			result: &Consumer{
-				Consumer: kong.Consumer{
-					Username: &username,
-				},
-			},
+			result:  &Consumer{Consumer: kong.Consumer{Username: &username}},
 			wantErr: true,
 		},
 		{
@@ -266,19 +234,11 @@ func TestConsumer_SetCredential(t *testing.T) {
 		{
 			name: "jwt without key",
 			args: args{
-				credType: "jwt",
-				consumer: &Consumer{
-					Consumer: kong.Consumer{
-						Username: &username,
-					},
-				},
+				credType:   "jwt",
+				consumer:   &Consumer{Consumer: kong.Consumer{Username: &username}},
 				credConfig: map[string]string{},
 			},
-			result: &Consumer{
-				Consumer: kong.Consumer{
-					Username: &username,
-				},
-			},
+			result:  &Consumer{Consumer: kong.Consumer{Username: &username}},
 			wantErr: true,
 		},
 		{
@@ -324,19 +284,11 @@ func TestConsumer_SetCredential(t *testing.T) {
 		{
 			name: "acl without group",
 			args: args{
-				credType: "acl",
-				consumer: &Consumer{
-					Consumer: kong.Consumer{
-						Username: &username,
-					},
-				},
+				credType:   "acl",
+				consumer:   &Consumer{Consumer: kong.Consumer{Username: &username}},
 				credConfig: map[string]string{},
 			},
-			result: &Consumer{
-				Consumer: kong.Consumer{
-					Username: &username,
-				},
-			},
+			result:  &Consumer{Consumer: kong.Consumer{Username: &username}},
 			wantErr: true,
 		},
 	}

--- a/internal/ingress/controller/parser/kongstate/consumer_test.go
+++ b/internal/ingress/controller/parser/kongstate/consumer_test.go
@@ -9,6 +9,7 @@ import (
 )
 
 func TestConsumer_SetCredential(t *testing.T) {
+	username := "example"
 	type args struct {
 		credType   string
 		consumer   *Consumer
@@ -47,6 +48,24 @@ func TestConsumer_SetCredential(t *testing.T) {
 			wantErr: false,
 		},
 		{
+			name: "key-auth without key",
+			args: args{
+				credType: "key-auth",
+				consumer: &Consumer{
+					Consumer: kong.Consumer{
+						Username: &username,
+					},
+				},
+				credConfig: map[string]string{},
+			},
+			result: &Consumer{
+				Consumer: kong.Consumer{
+					Username: &username,
+				},
+			},
+			wantErr: true,
+		},
+		{
 			name: "keyauth_credential",
 			args: args{
 				credType:   "keyauth_credential",
@@ -81,6 +100,24 @@ func TestConsumer_SetCredential(t *testing.T) {
 				},
 			},
 			wantErr: false,
+		},
+		{
+			name: "basic-auth without username",
+			args: args{
+				credType: "basic-auth",
+				consumer: &Consumer{
+					Consumer: kong.Consumer{
+						Username: &username,
+					},
+				},
+				credConfig: map[string]string{},
+			},
+			result: &Consumer{
+				Consumer: kong.Consumer{
+					Username: &username,
+				},
+			},
+			wantErr: true,
 		},
 		{
 			name: "basicauth_credential",
@@ -121,6 +158,24 @@ func TestConsumer_SetCredential(t *testing.T) {
 				},
 			},
 			wantErr: false,
+		},
+		{
+			name: "hmac-auth without username",
+			args: args{
+				credType: "hmac-auth",
+				consumer: &Consumer{
+					Consumer: kong.Consumer{
+						Username: &username,
+					},
+				},
+				credConfig: map[string]string{},
+			},
+			result: &Consumer{
+				Consumer: kong.Consumer{
+					Username: &username,
+				},
+			},
+			wantErr: true,
 		},
 		{
 			name: "hmacauth_credential",
@@ -167,6 +222,24 @@ func TestConsumer_SetCredential(t *testing.T) {
 			wantErr: false,
 		},
 		{
+			name: "oauth2 without client_id",
+			args: args{
+				credType: "oauth2",
+				consumer: &Consumer{
+					Consumer: kong.Consumer{
+						Username: &username,
+					},
+				},
+				credConfig: map[string]string{},
+			},
+			result: &Consumer{
+				Consumer: kong.Consumer{
+					Username: &username,
+				},
+			},
+			wantErr: true,
+		},
+		{
 			name: "jwt",
 			args: args{
 				credType: "jwt",
@@ -189,6 +262,24 @@ func TestConsumer_SetCredential(t *testing.T) {
 				},
 			},
 			wantErr: false,
+		},
+		{
+			name: "jwt without key",
+			args: args{
+				credType: "jwt",
+				consumer: &Consumer{
+					Consumer: kong.Consumer{
+						Username: &username,
+					},
+				},
+				credConfig: map[string]string{},
+			},
+			result: &Consumer{
+				Consumer: kong.Consumer{
+					Username: &username,
+				},
+			},
+			wantErr: true,
 		},
 		{
 			name: "jwt_secret",
@@ -229,6 +320,24 @@ func TestConsumer_SetCredential(t *testing.T) {
 				},
 			},
 			wantErr: false,
+		},
+		{
+			name: "acl without group",
+			args: args{
+				credType: "acl",
+				consumer: &Consumer{
+					Consumer: kong.Consumer{
+						Username: &username,
+					},
+				},
+				credConfig: map[string]string{},
+			},
+			result: &Consumer{
+				Consumer: kong.Consumer{
+					Username: &username,
+				},
+			},
+			wantErr: true,
 		},
 	}
 	for _, tt := range tests {

